contracts/drafts/ERC777/ERC777Base.sol

```diff
@@ -0,0 +1,491 @@
+pragma solidity ^0.5.2;
+
+import "./IERC777.sol";
+import "./IERC777TokensRecipient.sol";
+import "./IERC777TokensSender.sol";
+import "../../math/SafeMath.sol";
+import "../../utils/Address.sol";
+import "../../introspection/ERC1820Client.sol";
+
+/**
+ * @title ERC777 token implementation to inherit from
+ * @author etsvigun <utgarda@gmail.com>, Bertrand Masius <github@catageeks.tk>
+ * @dev see https://github.com/ethereum/EIPs/blob/master/EIPS/eip-777.md
+ * @dev inherit from this contract, do not use directly
+ */
+contract ERC777Base is IERC777, ERC1820Client {
+    using SafeMath for uint256;
+    using Address for address;
+
+    string private _name;
+
+    string private _symbol;
+
+    mapping(address => uint256) private _balances;
+
+    uint256 private _totalSupply;
+
+    uint256 private _granularity;
+
+    address[] private _defaultOpsArray;
+
+    bytes32 constant private SENDHASH = keccak256("ERC777TokensSender");
+    bytes32 constant private RECEIVEDHASH = keccak256("ERC777TokensRecipient");
+
+    mapping(address => bool) private _defaultOps;
+    mapping(address => mapping(address => bool)) private _revokedDefaultOps;
+    mapping(address => mapping(address => bool)) private _ops;
+
+    constructor(
+        string memory name,
+        string memory symbol,
+        uint256 granularity,
+        address[] memory defaultOperators
+    ) internal {
+        require(granularity > 0);
+        _name = name;
+        _symbol = symbol;
+        _granularity = granularity;
+        _defaultOpsArray = defaultOperators;
+        for (uint i = 0; i < defaultOperators.length; i++) {
+            _defaultOps[defaultOperators[i]] = true;
+        }
+        // register interface
+        setInterfaceImplementer(
+            address(this),
+            keccak256("ERC777Token"),
+            address(this)
+        );
+    }
+
+    /**
+    * @dev Send the amount of tokens from the address msg.sender to the address to
+    * @param to address recipient address
+    * @param amount uint256 amount of tokens to transfer
+    * @param data bytes information attached to the send, and intended for the recipient (to)
+     */
+    function send(
+        address to,
+        uint256 amount,
+        bytes calldata data
+    ) external {
+        _send(
```

Comment:
I think we should check here if `msg.sender` is allowed to do this, and have `_sender` be permissionless (same as how `ERC20`'s `transfer` and friends work).